sep: change to utilize cap-40 (#176)
stellar/experimental-payment-channels: specifications/sep-payment-channel-transactions.md

To setup the payment channel:

1. I creates escrow account EI.
2. R creates escrow account ER.
3. Set variable initial states:
   - s to EI's sequence number + 1.
   - i to 0.
   - e to 0.
4. Increment i.
5. I signs and shares:
   - A formation transaction F.
   - A declaration transaction D_i.
   - A closing transaction C_i, that closes the channel without any payment.
6. R signs and submits F.

Participants should defer deposits of initial contributions till after formation
for channels that will hold trustlines to issuers that are not auth immutable,
and could be clawback enabled. See [Security](#Security).

> Signatures for D, D_i, and C_i may be shared in a single message.

Reviewer: ```suggestion
Signatures for F, D_i, and C_i may be shared in a single message.
```

    not the native asset.
    - One `SET_OPTIONS` operation adjusting escrow account EI's thresholds such
    that I and R's signers must both sign.

>     - One `SET_OPTIONS` operation adding I signers to EI.

Reviewer: ❓ why is this only 1 now? I can only have 1 signer?

Author: Multiple signers are no longer supported because we are limited to two extra signers per transactions. This limits us to two subsequent transactions, and three total transactions, during the open process. This limits us to one subsequent transaction, and two total transactions, during the update process. There is an extra unused extra signers in the update process reserved for future use as a HTLC using the hash(x) signer.

I've added notes about this in the PR description. I should have included info about this.

  The `CHANGE_TRUST` operations configure the trustlines with the maximum limit,
  which is the maximum value of an `int64`, `0x7FFFFFFFFFFFFFFF`.

- D_i, with no payment, see [Payment](#Payment) process.

> - C_i, see [Payment](#Payment) process.

Reviewer: ❓  did you mean to switch these? C_i should have the no payment part

Author: Oops, fixed.

$22 to R.

To make a payment, participants agree on a new payment channel state. The
participant steps for which are:
participants:

1. Increment i.
2. Sign and exchange a closing transaction C_i.
3. Sign and exchange a declaration transaction D_i.
2. The payer participant signs and shares signatures for:
   - A declaration transaction D_i.
   - A closing transaction C_i.
3. The payee participant signs and shares signatures D_i and C_i.

> It is critical that signatures for D_i are exchanged after C_i because D_i will

Reviewer: awesome how much this simplifies stuff 😄

account EI and sequence number s_i+1, D_i leaves EI in a state where C_i can
execute.

  D_i has a single signer in its `extraSigners` precondition that ensures that
  if D_i is authorized, signed, and submitted, that the signature for C_i is
  revealed. The signer is specified as:

  - A ed25519 signed payload signer configured with:
    - Payload set to the transaction hash of C_i.

>     - Public key set to R's signer.

Reviewer: or `payee` might be more correct based off the above description
```suggestion
    - Public key set to other participant's signer.
```

Author: ```suggestion
    - Public key set to payee's signer.
```

1. Submit most recent D_i
2. Modify the most recent C_i `minSeqAge` and `minSeqLedgerGap` to zero
3. Resign and exchange the modified close transaction C_i
4. Submit modified C_i

If participants choose to coordinate a close before submitting D_i they must
take care to also modify D_i such that it contains the new C_i's hash is
included in the new D_i's `extraSigners` or to keep a copy of the old C_i's

> signatures required to satisfy D_i's `extraSigners` precondition.

Author: @acharb I added this warning to the SEP following our conversation.

Reviewer: 👍 . I think there's a typo though, should be?:
```
If participants choose to coordinate a close before submitting D_i they must
take care that the new C_i's hash is included in the D_i's `extraSigners`,
or keep a copy of the old C_i's signatures required to satisfy D_i's `extraSigners` 
precondition.
```

Author: ```suggestion
If participants choose to coordinate a close before submitting D_i they must
take care that the new C_i's hash is included in the D_i's `extraSigners`,
or keep a copy of the old C_i's signature required to satisfy D_i's `extraSigners` 
precondition.
```

1. Increment i.
2. I and R build the bump transaction B_i.
3. Set e to i.
4. I or R signs and shares signatures for:
   - A declaration transaction D_i+1.
   - A closing transaction C_i+1, that closes the channel with disbursement
   matching the most recent agreed state.
5. R or I signs and shares signatures for B_i, D_i+1, and C_i+1.
6. Increment i.
7. I or R submit B_i.

The transactions are constructed as follows:

- B_i, the _bump transaction_, bumps the sequence number of escrow account E
such that only the most recent transaction set is valid. B has source account
EI, sequence number s_i.

  B_i has two signers in its `extraSigners` precondition that ensures that if
  B_i is authorized, signed, and submitted, that the signatures for D_i+1 and
  C_i+1 are revealed. The signer is specified as:

  - A ed25519 signed payload signer configured with:
    - Payload set to the transaction hash of D_i.

>     - Public key set to R's signer.

Reviewer: ❓ same thing here, this should be `other participant's signer`?

Author: Other's participants here would be ambigious because both participants build the transaction and they need to use the same participant here. You're right though, it isn't always R here. I've updated this section in 8c35f26.

deposit after formation, unless they trust the asset issuer not to clawback from
payment channel escrow accounts, or unless the asset is auth immutable.

## Transaction Signature Disclosure

> Processes that require the signing of multiple transactions make use an ed25519

Reviewer: ```suggestion
Processes that require the signing of multiple transactions make use of an ed25519
```
